test/distributed/test_compute_comm_reordering.py

```diff
@@ -0,0 +1,267 @@
+from torch.fx.experimental.proxy_tensor import make_fx
+from torch.testing._internal.common_distributed import (
+    DynamoDistributedSingleProcTestCase,
+    DynamoDistributedMultiProcTestCase,
+    _dynamo_dist_per_rank_init,
+    requires_nccl,
+    skip_if_lt_x_gpu,
+)
+from torch._inductor.compile_fx import compile_fx as inductor_compile_fx
+from torch._inductor.utils import has_triton, run_and_get_triton_code
+import torch._dynamo.logging
+from torch._inductor import ir
+
+
+
+@requires_nccl()
+class TestComputeCommReorderingMultiProc(DynamoDistributedMultiProcTestCase):
+    """
+    Run correctness checks in multi-proc runner, mark with minimum # GPUs to run under
+    """
+    def get_world_trs(self):
+        return {
+            "tag": "",
+            "ranks": list(range(self.world_size)),
+            "group_size": self.world_size,
+        }
+
+    @property
+    def world_size(self) -> int:
+        # hack: no matter whether we have 2 or 3 or 4 gpus, just run on 2
+        # works around issue with skipif<2 and workers with unpredictable #s gpu
+        return 2
+
+    """
+    TODO: unit tests to add:
+    1. sink_waits()
+        TODO: design good # and variaty of compute ops
+        TODO: assert correct ordering
+    2. raise_comms()
+        TODO: design good # and variaty of compute ops
+        TODO: assert correct ordering
+    3. reorder_compute_for_overlap()
+        1) only 1 comm
+        2) 2+ comms
+        TODO: design good # and variaty of compute ops
+        TODO: assert correct ordering
+    4. reorder_compute_and_comm_for_overlap() integration test
+        TODO: assert correct ordering
+    """
+
+    @unittest.skipIf(not has_triton(), "Inductor+gpu needs triton and recent GPU arch")
+    @skip_if_lt_x_gpu(2)
+    @patch.object(torch._inductor.config, "allow_buffer_reuse", True)
+    # TODO: somehow inductor bg compile threads are causing hangs at exit with distributed work dtor
+    @patch.object(torch._inductor.config, "compile_threads", 1)
+    @patch.object(torch._inductor.config, "reorder_for_compute_comm_overlap", True)
+    @patch.object(torch._inductor.config, "reorder_for_compute_comm_overlap_passes", [
+        "sink_waits",
+    ])
+    def test_sink_waits(self):
+        def func(a, *, tag, ranks, group_size):
+            ar = _functional_collectives.all_reduce(a, "sum", ranks, tag)
+            c = torch.relu(a)
+            d = torch.matmul(c, c)
+            e = d + ar
+            return (e,)
+
+        with _dynamo_dist_per_rank_init(self.rank, self.world_size):
+            inputs = torch.ones(4, 4, dtype=torch.float, device="cuda") + self.rank
+            compiled = torch.compile(func)
+            code = run_and_get_triton_code(compiled, inputs, **self.get_world_trs())
+            # NOTE: notice that `_wait_tensor` is delayed until right before first use
+            FileCheck() \
+                .check("buf0 = empty_strided((4, 4), (4, 1), device='cuda', dtype=torch.float32)") \
```

Comment:
I'm not a big fan of these FileCheck tests, matching against the entire generated output. I think there's too much churn from random things that get changed, and updating these tests will be quite painful.

One option is to use `assertExpectedEqual`, which at least provides a more automated way of optimizing it.

Another thing we can do is after our scheduling, we can write a simple simulator based off of our nodes' estimated runtime to compute the total runtime and things like "exposed comms".

Reply:
That would also allow us to potentially run these tests without actually needing multiple GPUs, which would be a huge win I think.

Reply:
> That would also allow us to potentially run these tests without actually needing multiple GPUs, which would be a huge win I think.

couldn't we run a test with collectives in it on a single gpu today? the collective should still work at all layers even if its essentially a no-op.

> I'm not a big fan of these FileCheck tests

its definitely fragile to match too much of the generated code.  especially variable names.. but unless we actually build some more infra for validation of certain properties of IR its the best we have.  

What does assertExpectedEqual do that is helpful here?

Reply:
> What does assertExpectedEqual do that is helpful here?

It allows us to automatically update all test files that have jitter in the output.

> couldn't we run a test with collectives in it on a single gpu today? the collective should still work at all layers even if its essentially a no-op.

Does this work? Even through Inductor?

Reply:
> Does this work? Even through Inductor? Well regardless, we'd still want property-based testing.

well.. it definitely works in eager. i don't see why it wouldn't work in inductor but i didn't try it

Reply:
I'm wondering whether it gets caught and dispatched to a clone somewhere.